default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1602 @@
+                ch = Cache.production_cost[self.pid][self.hull.name]
+                p1 = a*s*ca + a*ch
+                p2 = math.sqrt(a * (ca*s + ch) * (a*s*cw+a*ch+h*cw-h*ca))
+                p3 = a*(ca-cw)
+                n = max((p1+p2)/p3, (p1-p2)/p3)
+                n = int(round(n))
+                n = max(n, 1)
+                n = min(n, s)
+                print "estimated weapon slots for %s: %d" % (self.hull.name, n)
+                ret_val[idxarmour] = s-n
+                ret_val[idxweapon] = n
+            else:
+                ret_val[idxweapon] = num_slots
+        elif armours:
+            armour = max(armours, key=cap).name
+            idxarmour = available_parts.index(armour)
+            ret_val[idxarmour] = num_slots
+        else:
+            ret_val[-1] = num_slots
+        return ret_val
+
+
+class AIShipDesign_Trooper(AIShipDesign):
+    """Base class for troop ships. To be inherited from.
+
+    Extends __init__()
+    Overrides _rating_function()
+    Overrides _starting_guess()
+    Overrides _class_specific_filter
+    """
+    basename = "Troopers (Do not build me)"
+    description = "Trooper."
+    useful_part_classes = TROOPS
+    filter_useful_parts = True
+    filter_inefficient_parts = True
+    running_index = 1
+
+    def __init__(self):
+        AIShipDesign.__init__(self)
+        self.additional_specifications.minimum_structure = self.additional_specifications.enemy_mine_dmg * 2
+
+    def _rating_function(self):
+        if self.troops == 0:
+            return -99999
+        else:
+            return self.troops/self.production_cost
+
+    def _starting_guess(self, available_parts, num_slots):
+        # fill completely with biggest troop pods. If none are available for this slot type, leave empty.
+        troop_pods = [_get_part_type(part) for part in available_parts if _get_part_type(part).partClass in TROOPS]
+        ret_val = (len(available_parts)+1)*[0]
+        if troop_pods:
+            cap = lambda x: x.capacity
+            biggest_troop_pod = max(troop_pods, key=cap).name
+            try:  # we could use an if-check here but since we usually have troop pods for the slot, try is faster
+                idx = available_parts.index(biggest_troop_pod)
+            except ValueError:
+                idx = len(available_parts)
+                traceback.print_exc()
+        else:
+            idx = len(available_parts)
+        ret_val[idx] = num_slots
+        return ret_val
+
+    def _class_specific_filter(self, partname_dict):
+        for slot in partname_dict:
+            remaining_parts = [part for part in partname_dict[slot] if _get_part_type(part).partClass in TROOPS]
+            partname_dict[slot] = remaining_parts
+
+
+class AIShipDesign_Trooper_Orbital(AIShipDesign_Trooper):
+    """Class implementing orbital invasion designs
+
+    Extends __init__()
+    """
+    basename = "SpaceInvaders"
+    description = "Ship designed for local invasions of e
```

Comment:
The new module is called in the current ```generateProductionOrders()``` in ProductionAI which already had references to ```ColonisationAI.empire_shipyards``` so this should be fine.

However, it might make sense to check for this in the future.